### ISSUE TITLE
Critical dependency: require function is used in a way in which dependencies cannot be statically extracted

### ISSUE THREAD

**Issue by pgg** [vladmandic/human#209]
**Issue Description**
```import { Human } from '@vladmandic/human';``` emits ```./node_modules/@vladmandic/human/dist/human.esm.js:25:70-77 - Warning: Critical dependency: require function is used in a way in which dependencies cannot be statically extracted``` for latest angular project and stackblitz asks to install ```@tensorflow/tfjs-node```

**Steps to Reproduce**
https://stackblitz.com/edit/angular-ivy-sagwez?file=src/app/app.component.ts

**Environment**

- Human library version? 2.5.1
- Built-in demo or custom code? custom
- Type of module used (e.g. `js`, `esm`, `esm-nobundle`)? esm
- Browser or NodeJS and version (e.g. *NodeJS 14.15* or *Chrome 89*)? chrome 95
- OS and Hardware platform (e.g. *Windows 10*, *Ubuntu Linux on x64*, *Android 10*)? macos
- Framework (if any) (e.g. *React*, *NextJS*, etc.)? angular

**Comment by vladmandic** [comment]
Seems that by default, packager for angular picks up `NodeJS` version of `Human` so it fails when building the bundle  
Simply change import statement to point to exact version of the library and its ok  

```js
import { Human } from '@vladmandic/human/dist/human.esm.js';
```

**Comment by pgg** [comment]
works though vscode complains with: Could not find a declaration file for module '@vladmandic/human/dist/human.esm.js'. '.node_modules/@vladmandic/human/dist/human.esm.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/vladmandic__human` if it exists or add a new declaration (.d.ts) file containing `declare module '@vladmandic/human/dist/human.esm.js';`ts(7016)

**Comment by vladmandic** [comment]
yeah, automatically mapping of typedefs only works for package imports, not named imports  
ive created explicit typedef few days ago to handle that scenario - its in main branch  

will be released as part of 2.5.2 (next week)  
or you can install `human` from **git** directly (`npm install git+https://github.com/vladmandic/human`) instead of **npmjs**

**Comment by pgg** [comment]
ok thanks

**Comment by pgg** [comment]
> Seems that by default, packager for angular picks up `NodeJS` version of `Human`, so it fails when building the bundle for the web
> 
> Simply change import statement to point to exact version of the library and its ok
> 
> ```js
> import { Human } from '@vladmandic/human/dist/human.esm.js';
> ```

this seems to only be the case with stackblitz and not the local angular cli

**Comment by pgg** [comment]
both: ```// import Human from '@vladmandic/human';``` and ```import { Human } from '@vladmandic/human/dist/human.esm.js';``` still emits ```./node_modules/@vladmandic/human/dist/human.esm.js:25:70-77 - Warning: Critical dependency: require function is used in a way in which dependencies cannot be statically extracted``` in local angular build for v 2.5.2

**Comment by vladmandic** [comment]
wait, you said this only happens in **stackblitz** build, not **local** build?  
can you share a full reproduction that i can start on my system?

**Comment by pgg** [comment]
> wait, you said this only happens in **stackblitz** build, not **local** build? can you share a full reproduction that i can start on my system?

https://github.com/pgg/h209

**Comment by vladmandic** [comment]
# TLDR

fully working with new typedefs and tfjs bundle and angular/webpack are no longer complaining

# long version

refactored how typedefs are generated
- extract typedefs from tfjs
- compile typedefs for human
- create single rollup `human.d.ts` using `api-extractor`

refactored build for `@vladmandic/tfjs`
- detect `nodejs` specific imports and exclude references from bundle
- add missing types for `OffscreenCanvas` and `Long`

## webpack test

> `webpack.config.js`
```js
module.exports = {
  mode: 'none',
  target: ['web'],
  entry: 'index.js',
  stats: { errorDetails: true },
  output: { path: '/home/vlado/dev/webpack/dist', filename: 'index.js' },
};
```

> `index.js`
```
import { Human } from '@vladmandic/human';
const human = new Human();
console.log(human.version, human.tf.version);
console.log();
human.init();
```

    # webpack
    asset index.js 2.84 MiB [emitted] (name: main)
    runtime modules 921 bytes 5 modules
    cacheable modules 2.83 MiB
      ./src/index.js 328 bytes [built] [code generated]
      ../human/dist/human.esm.js 2.83 MiB [built] [code generated]
    webpack 5.64.1 compiled successfully in 1873 ms

## angular test

    git clone https://github.com/pgg/h209
    npm install
    npm uninstall @vladmandic/human
    npm install git+https://github.com/vladmandic/human
    npm build

    Initial Chunk Files           | Names         |     Size
    main.cfc54ecb2295d37e.js      | main          |  1.66 MB
    polyfills.6ece23527d8a58cc.js | polyfills     | 36.19 kB
    runtime.be13367bb86a04b1.js   | runtime       |  1.05 kB
    styles.ef46db3751d8e999.css   | styles        |  0 bytes
                                  | Initial Total |  1.70 MB
    Build at: 2021-11-18T14:59:08.898Z - Hash: 1e668dff8fd9cbce - Time: 3037ms

**Comment by pgg** [comment]
see updated https://github.com/pgg/h209 ```Error: node_modules/@vladmandic/human/types/human.d.ts:1:23 - error TS2688: Cannot find type definition file for 'offscreencanvas'.```

**Comment by vladmandic** [comment]
i've just bundled `@types/offscreencanvas` in `human.d.ts`, version 2.5.4 will be published soon.

**Comment by pgg** [comment]
fixed in 2.5.4

**Comment by gcardacci** [comment]
Hello, first of all thank you for the library!

I have installed version 2.11.1 but I still have the same warning.
I get two warnings from the same file:


```
WARNING in ./node_modules/@vladmandic/human/dist/human.esm.js 9:73-80
Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
```

and

```
WARNING in ./node_modules/@vladmandic/human/dist/human.esm.js 13:11-18
Critical dependency: require function is used in a way in which dependencies cannot be statically extracted
```

I have inspected the file "human.esm.js" and I can see the use of the word "require" inside it, It may be interpreted as a reserved word.

![image](https://user-images.githubusercontent.com/19553612/200064891-26362e1a-b2e5-46a2-9573-8965b8b94669.png)

**Comment by vladmandic** [comment]
@gcardacci if you have an issue it may not be the same so please open a new issue  
as this one is marked as resolved for over a year  

and note exactly in which environment it is happening and what does your project look like  
and if you can provide a reproduction, it would help a lot
